### PR TITLE
BUG: Fix rolling functions with variable windows on decreasing index

### LINKED ISSUE #32385: BUG: Rolling functions doesn't work on decreasing variable index
#### Code Sample

```python
    def test_rolling_on_decreasing_index(self):
        # GH-19248
        index = [
            Timestamp("20190101 09:00:00"),
            Timestamp("20190101 09:00:02"),
            Timestamp("20190101 09:00:03"),
            Timestamp("20190101 09:00:05"),
            Timestamp("20190101 09:00:06"),
        ]

        df = DataFrame({"column": [3, 4, 4, 2, 1]}, index=reversed(index))
        result = df.rolling("2s").min()
        expected = DataFrame(
            {"column": [3.0, 4.0, 4.0, 2.0, 1.0]}, index=reversed(index)
        )
        # actual incorrect output is [3, 3, 3, 2, 1]
        tm.assert_frame_equal(result, expected)
```
#### Problem description

Rolling window operations on variable decreasing index don't work even after https://github.com/pandas-dev/pandas/pull/28297. The reason for this is that the `calculate_variable_window_bounds` function doesn't work with decreasing index. Tested with latest master (~version 1.0.1).

### PR DESCRIPTION
**Pull request by leftys**
- [x] Closes https://github.com/pandas-dev/pandas/issues/32385
- [x] tests updated & passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/window/test_timeseries_window.py [hunk 1]
```diff
@@ -721,7 +721,7 @@ def test_rolling_on_decreasing_index(self):
         df = DataFrame({"column": [3, 4, 4, 2, 1]}, index=reversed(index))
         result = df.rolling("2s").min()
         expected = DataFrame(
-            {"column": [3.0, 3.0, 3.0, 2.0, 1.0]}, index=reversed(index)
+            {"column": [3.0, 4.0, 4.0, 2.0, 1.0]}, index=reversed(index)
```
Comment (leftys) [inline pandas/tests/window/test_timeseries_window.py]: The original expected value was wrong: https://github.com/pandas-dev/pandas/pull/28297#discussion_r351669004.

File: pandas/tests/window/test_timeseries_window.py [hunk 2]
```diff
@@ -709,19 +709,19 @@ def test_rolling_cov_offset(self):
...
+            Timestamp("20190101 09:00:10"),
         ]
 
-        df = DataFrame({"column": [3, 4, 4, 2, 1]}, index=reversed(index))
-        result = df.rolling("2s").min()
+        df = DataFrame({"column": [3, 4, 4, 5, 6]}, index=index)
+        result = df.rolling("5s").min()
         expected = DataFrame(
-            {"column": [3.0, 3.0, 3.0, 2.0, 1.0]}, index=reversed(index)
+            {"column": [3.0, 3.0, 4.0, 4.0, 6.0]}, index=index
```
Comment (leftys) [inline pandas/tests/window/test_timeseries_window.py]: The original expected value was wrong: https://github.com/pandas-dev/pandas/pull/28297#discussion_r351669004.

File: doc/source/whatsnew/v1.0.2.rst [hunk 3]
```diff
@@ -83,6 +83,10 @@ Bug fixes
 - Fixed bug where :meth:`GroupBy.first` and :meth:`GroupBy.last` would raise a ``TypeError`` when groups contained ``pd.NA`` in a column of object dtype (:issue:`32123`)
 - Fix bug in :meth:`Series.convert_dtypes` for series with mix of integers and strings (:issue:`32117`)
 
+**Rolling**
+
+- Fix bug in :meth:`calculate_variable_window_bounds` that couldn't work on decreasing index (:issue:`32385`).
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.2.rst]: Can you describe this in terms of the public API? (so `rolling()` with time frequency, I suppose)

**Comment by jorisvandenbossche** [comment]
cc @mroeschke are you familiar with this code? 

I can confirm that indeed the current output (and tested behaviour) is wrong

**Comment by mroeschke** [comment]
Only familiar from a refactor standpoint. I didn't write the original window bounds calculations.

File: pandas/_libs/window/indexers.pyx [hunk 4]
```diff
@@ -44,6 +44,7 @@ def calculate_variable_window_bounds(
     cdef:
         bint left_closed = False
         bint right_closed = False
+        int index_growth_sign = +1
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx]: `+1` -> `1`

File: pandas/_libs/window/indexers.pyx [hunk 5]
```diff
@@ -58,6 +59,9 @@ def calculate_variable_window_bounds(
     if closed in ['left', 'both']:
         left_closed = True
 
+    if index[num_values-1] < index[0]:
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx]: `index[num_values-1]` -> `index[num_values - 1]`

File: pandas/tests/window/test_timeseries_window.py [hunk 6]
```diff
@@ -709,20 +709,18 @@ def test_rolling_cov_offset(self):
...
-            Timestamp("20190101 09:00:05"),
-            Timestamp("20190101 09:00:06"),
+            Timestamp("20190101 09:00:30"),
+            Timestamp("20190101 09:00:27"),
+            Timestamp("20190101 09:00:20"),
+            Timestamp("20190101 09:00:18"),
+            Timestamp("20190101 09:00:10"),
         ]
 
-        df = DataFrame({"column": [3, 4, 4, 2, 1]}, index=reversed(index))
```
Comment (WillAyd) [inline pandas/tests/window/test_timeseries_window.py]: Wouldn't this have been working previously with a decreasing index?

Comment (jorisvandenbossche) [reply pandas/tests/window/test_timeseries_window.py]: Not fully sure what the question is, but: on 0.25 we raised an error for decreasing index. For 1.0, rolling with a decreasing index was enabled, but gave wrong output (see eg https://github.com/pandas-dev/pandas/issues/32385 for what the output of this should actually be)

Comment (leftys) [reply pandas/tests/window/test_timeseries_window.py]: 1.0 would return [3,3,3,2,1] for this case. The rolling window covered the whole series.

File: doc/source/whatsnew/v1.0.2.rst (hunk 3 above)
Comment (leftys) [reply doc/source/whatsnew/v1.0.2.rst]: updated

File: pandas/_libs/window/indexers.pyx (hunk 4 above)
Comment (leftys) [reply pandas/_libs/window/indexers.pyx]: done

File: pandas/_libs/window/indexers.pyx (hunk 5 above)
Comment (leftys) [reply pandas/_libs/window/indexers.pyx]: done

File: pandas/_libs/window/indexers.pyx [hunk 7]
```diff
@@ -58,6 +59,9 @@ def calculate_variable_window_bounds(
     if closed in ['left', 'both']:
         left_closed = True
 
+    if index[num_values - 1] < index[0]:
```
Comment (jreback) [inline pandas/_libs/window/indexers.pyx]: are we checking if len > 0 before this?

File: pandas/tests/window/test_timeseries_window.py [hunk 8]
```diff
@@ -709,20 +709,18 @@ def test_rolling_cov_offset(self):
         tm.assert_series_equal(result, expected2)
 
     def test_rolling_on_decreasing_index(self):
-        # GH-19248
+        # GH-19248, GH-32385
```
Comment (jreback) [inline pandas/tests/window/test_timeseries_window.py]: can you confrm that we have tests for min/max on an empty frame

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. a question.

**Comment by TomAugspurger** [comment]
@leftys can you respond to https://github.com/pandas-dev/pandas/pull/32386#pullrequestreview-368778011? Trying to get 1.0.2 out soon.

File: pandas/_libs/window/indexers.pyx (hunk 7 above)
Comment (leftys) [reply pandas/_libs/window/indexers.pyx]: i guess this is done in some caller method, as we access `start[0]` and start is initialized to the same length as `index`. (checked by a test too)

File: pandas/tests/window/test_timeseries_window.py (hunk 8 above)
Comment (leftys) [reply pandas/tests/window/test_timeseries_window.py]: test added

**Comment by leftys** [comment]
> @leftys can you respond to [#32386 (review)](https://github.com/pandas-dev/pandas/pull/32386#pullrequestreview-368778011)? Trying to get 1.0.2 out soon.

@TomAugspurger done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @leftys very nice!

**Comment by jreback** [comment]
@meeseeksdev backport 1.0.x

**Pull request merged**